File: graphql/src/service/post.js
import rp from 'request-promise';

const POST_API_URL = 'http://localhost:8080/api/posts';

export const getAllPosts = () => {
    return rp.get(POST_API_URL, {
        json: true
    });
};


export const getPostById = (id) => {
    return rp.get(`${POST_API_URL}/${id}`, {
        json: true
    });
};

export const getPostsByAuthor = (authorID) => {
    return rp.get(`${POST_API_URL}/search?author=${authorID}`, {
        json: true
    });
};

export default {
    getAllPosts,
    getPostById,
    getPostsByAuthor,
};

File: go/src/posts-service/core/api/post.go
package api

import (
	"encoding/json"
	"fmt"
	"github.com/gorilla/mux"
	"net/http"
	"posts-service/core/service"
)

func ListHandler(service service.Post) func(w http.ResponseWriter, r *http.Request) {
	return func(w http.ResponseWriter, r *http.Request) {
		w.Header().Add("content-type", "application/json")
		posts := service.GetAll()

		if err := json.NewEncoder(w).Encode(posts); err != nil {
			fmt.Printf("error")
			w.WriteHeader(http.StatusInternalServerError)
			return
		}
	}
}

func PostHandler(service service.Post) func(w http.ResponseWriter, r *http.Request) {
	return func(w http.ResponseWriter, r *http.Request) {
		w.Header().Add("content-type", "application/json")
		id := mux.Vars(r)["id"]
		post := service.GetById(id)

		if err := json.NewEncoder(w).Encode(post); err != nil {
			fmt.Printf("error")
			w.WriteHeader(http.StatusInternalServerError)
			return
		}
	}
}

func PostByAuthorHandler(service service.Post) func(w http.ResponseWriter, r *http.Request) {
	return func(w http.ResponseWriter, r *http.Request) {
		w.Header().Add("content-type", "application/json")
		authorID := r.URL.Query().Get("author")
		posts := service.GetByAuthorId(authorID)

		if err := json.NewEncoder(w).Encode(posts); err != nil {
			fmt.Printf("error")
			w.WriteHeader(http.StatusInternalServerError)
			return
		}
	}
}